File: Fortran/2_pingPong/c/pingPong.f90
program pingPong
    use mpi
    implicit none

    ! Maximum array size 2^10 = 1024 elements
    integer, parameter :: MAX_ARRAY_SIZE = 2**10

    ! Variables for the process rank and number of processes
    integer :: myRank, numProcs, ierr
    integer, allocatable, dimension(:) :: myArray
    integer :: numberOfElementsToSend, numberOfElementsReceived

    ! Variables to generate random number
    real(KIND=4) :: randomReal

    ! Initialize MPI
    call MPI_INIT(ierr)
    ! Find out MPI communicator size and process rank
    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    ! PART B
    call random_seed()
    call random_number(randomReal)
    numberOfElementsToSend =  floor( randomReal * 100.0 )

    ! Allocate an array big enough to hold even the largest message
    allocate(myArray(MAX_ARRAY_SIZE))
    if (.not.allocated(myArray)) then
        print *,"Not enough memory"
        stop 1
    end if

    ! TODO: Display an error message if too few processes

    ! Have only the first process execute the following code
    if (myRank == 0) then
        print '("Sending ",I0," elements")', numberOfElementsToSend
        ! Send numberOfElementsToSend and the elements themselves
        call MPI_SEND(numberOfElementsToSend, 1, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, ierr)
        call MPI_SEND(myArray, numberOfElementsToSend, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, ierr)
        ! Receive elements count and array elements
        call MPI_RECV(numberOfElementsReceived, 1, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
        call MPI_RECV(myArray, numberOfElementsReceived, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

        print '("Received ",I0," elements")', numberOfElementsReceived
    else ! myRank == 1
        ! TODO: Make it work with more than two processes
        ! Receive the ping and send a pong
        call MPI_RECV(numberOfElementsReceived, 1, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
        call MPI_RECV(myArray, numberOfElementsReceived, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

        print '("Received ",I0," elements")', numberOfElementsReceived

        print '("Sending back ",I0," elements")', numberOfElementsToSend

        call MPI_SEND(numberOfElementsToSend, 1, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, ierr)
        call MPI_SEND(myArray, numberOfElementsToSend, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, ierr)
    end if

    deallocate(myArray)

    ! Finalize MPI
    call MPI_FINALIZE(ierr)
end program pingPong

File: C/2_pingPong/solution/b/pingPong.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

// Maximum array size 2^10 = 1024 elements
#define MAX_ARRAY_SIZE (1<<10)

int main(int argc, char **argv)
{
    // Variables for the process rank and number of processes
    int worldRank, worldSize;

    // Initialize MPI
    MPI_Init(&argc, &argv);
    // Find out MPI communicator size and process rank
    MPI_Comm_size(MPI_COMM_WORLD, &worldSize);
    MPI_Comm_rank(MPI_COMM_WORLD, &worldRank);

    // PART B
    srand(MPI_Wtime()*100000 + worldRank*137);
    int numberOfElementsToSend = rand() % 100;
    // Allocate an array big enough to hold the largest message in our set
    int *myArray = (int *)malloc(sizeof(int)*MAX_ARRAY_SIZE);
    if (myArray == NULL)
    {
        printf("Not enough memory\n");
        exit(1);
    }
    int numberOfElementsReceived;

    // Have only the first process execute the following code
    if (worldRank == 0)
    {
        printf("Sending %i elements\n", numberOfElementsToSend);
        // Send "numberOfElementsToSend" elements (ping)
        // -- first send a message with the number of elements
        MPI_Send(&numberOfElementsToSend, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
        // -- send the elements themselves
        MPI_Send(myArray, numberOfElementsToSend, MPI_INT, 1, 0,
            MPI_COMM_WORLD);
        // Receive elements (pong)
        // Store number of elements received in numberOfElementsReceived
        // -- first message brings the number of elements
        MPI_Recv(&numberOfElementsReceived, 1, MPI_INT, 1, 0, MPI_COMM_WORLD,
            MPI_STATUS_IGNORE);
        // -- receive the elements themselves
        MPI_Recv(myArray, numberOfElementsReceived, MPI_INT, 1, 0,
            MPI_COMM_WORLD, MPI_STATUS_IGNORE);

        printf("Received %i elements\n", numberOfElementsReceived);
    }
    else // worldRank == 1
    {
        // Receive elements (in two steps - see comments above)
        // Store number of elements received in numberOfElementsReceived
        MPI_Recv(&numberOfElementsReceived, 1, MPI_INT, 0, 0, MPI_COMM_WORLD,
            MPI_STATUS_IGNORE);
        MPI_Recv(myArray, numberOfElementsReceived, MPI_INT, 0, 0,
            MPI_COMM_WORLD, MPI_STATUS_IGNORE);

        printf("Received %i elements\n", numberOfElementsReceived);

        printf("Sending back %i elements\n", numberOfElementsToSend);
        // Send "numberOfElementsToSend" elements
        MPI_Send(&numberOfElementsToSend, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
        MPI_Send(myArray, numberOfElementsToSend, MPI_INT, 0, 0,
            MPI_COMM_WORLD);
    }

    // Finalize MPI
    MPI_Finalize();

    return 0;
}
